File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int function(int[] weights,int capacity){
        int days = 1,load=0;
        for(int w:weights){
            if(load+w>capacity){
                days = days+1;
                load = w;
            }
            else{
                load+=w;
            }
        }
        return days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int max = weights[0],sum=weights[0];
        for(int i=1;i<weights.length;i++){
            if(weights[i]>max) max = weights[i];
            sum+=weights[i];
        }
        int low = max,high = sum;
        while(low<=high){
            int mid = (low+high)/2; //capacity
            int numDays = function(weights,mid);
            if(numDays<=days){
                high = mid-1;
            }
            else{
                low = mid+1;
            }
        }
        return low;
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def daysRequired(self,weights,capacity):
        days,curSum = 1,0
        for w in weights:
            if curSum+w<=capacity:
                curSum+=w
            else:
                days+=1
                curSum = w
        return days
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low,high = max(weights),sum(weights)
        while low<=high:
            mid = low + (high-low)//2
            daysNeeded = self.daysRequired(weights,mid)
            if daysNeeded<=days:
                high = mid-1
            else:
                low = mid+1
        return low